controllers/helmchart_controller.go

```diff
@@ -621,6 +647,17 @@ func (r *HelmChartReconciler) buildFromHelmRepository(ctx context.Context, obj *
 	opts := chart.BuildOptions{
 		ValuesFiles: obj.GetValuesFiles(),
 		Force:       obj.Generation != obj.Status.ObservedGeneration,
+		// The remote builder will not attempt to download the chart if
+		// an artifact exists with the same name and version and `Force` is false.
+		// It will try to verify the chart if:
+		// - we are on the first reconciliation
+		// - the HelmChart spec has changed (generation drift)
+		// - the previous reconciliation resulted in a failed artifact verification
```

Comment:
Not sure if this would be true. I was going to suggest to remove SourceVerified=False condition if present at the beginning of `reconcileSource()`, similar to OCIRepo https://github.com/fluxcd/source-controller/blob/35ea086358793a706cd09ef0b0855191a3e4e5e7/controllers/ocirepository_controller.go#L307-L313, due of their negative value to be of highest priority that'll overshadow every other failure condition which may be the main reason for a failure.
So, instead of saying "failed artifact verification", more accurate would be "unverified".
I see that this same comment exists in OCIRepo reconciler, which also should be inaccurate.

Reply:
I've not tested this myself yet, only theory. Please verify that this is true by causing a verification failure and then a fetch failure in a subsequent reconciliation. Verification failure reason and message would appear in the Ready condition instead of Fetch Failed reason and message.

Reply:
We do not have access to the digest, so we cannot make the call to verify based on that. If we remove the failure condition, we won't be able to verify at next reconciliation. The digest on the other hand might have been changed by the user upon realising that a verification attempt change.

Reply:
Just noticed that the reason above is `ChartVerificationError` and not `VerificationError` that GitRepo and OCIRepo have. But this is the value of `BuildFailed` condition, not `SourceVerified` condition, so maybe intentional, nothing to do?

If I'm not misreading the pasted output, looks like you've created BuildFailed=True, not SourceVerified=False.
Going through the code change again, looks like SourceVerified=False condition is only set when `makeVerifiers()` returns an error. So, if there's no verifier because the secret doesn't contain any public key, SourceVerified=False. The verification can not be performed at all. But the code below checks if SourceVerified=False, chart build should perform verification, which wouldn't work, as there's no verifier to verify. 

In order to recreate the scenario that I was talking about initially, put a non-existing secret reference for verify and let SourceVerified=False be added in the status. And then maybe delete the HelmRepository secret so that there's an auth failure, resulting in FetchFailed=True condition. In this situation, Ready should surface FetchFailed reason and not SourceVerified reason in Ready condition value.

In other words, SourceVerifiedCondition when False should be recalculated and not be reused from previous reconciliation failure. As described in GitRepo and OCIRepo's `reconcileSource()` code, SourceVerifiedCondition is a bipolarity condition and has the highest priority compared to everything else when it's False and it should not be retained in the status in its negative value.

Coming back to `chart.BuildOptions.Verify` below, I think it's difficult to decide whether to verify or not here, because as you said, we don't have access to the digest or enough information about the chart. So, maybe here, we can mark Verify to be true if obj.Spec.Verify is not nil. I think a SourceVerified=False if happens above would not continue the reconciliation and fail before reaching here. Maybe somewhere in the chart build process, we need to determine if a re-verification is needed.

Reply:
Thanks @darkowlzz I have reproduced what you said.

> So, maybe here, we can mark Verify to be true if obj.Spec.Verify is not nil. I think a SourceVerified=False if happens above would not continue the reconciliation and fail before reaching here.

I don't see any other choice. This is not the best performance wise as we would most of the time keep reverifying the same chart.

> Maybe somewhere in the chart build process, we need to determine if a re-verification is needed. 

We will have to ask the helm folks to add the digest to the chart metadata... I will talk about this with @scottrigby

Reply:
> Just noticed that the reason above is ChartVerificationError and not VerificationError that GitRepo and OCIRepo have. But this is the value of BuildFailed condition, not SourceVerified condition, so maybe intentional, nothing to do?

The verification is an option enabled as part of the build, so it seems right to report the verification failure as part of the build failure.